business/apps.go

```diff
@@ -490,29 +488,61 @@ func (in *AppService) fetchNamespaceApps(ctx context.Context, namespace string,
 		keys = append(keys, k)
 	}
 	slices.Sort(keys)
+
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+	errChan := make(chan error, len(keys)) // for all keys
+
 	allEntities := make(namespaceApps)
+
 	for _, k := range keys {
 		w := ws[k]
-		// WorkloadGroup.Labels can be empty
-		if len(w.Labels) > 0 {
-			// Check if namespace is cached
-			serviceCriteria := ServiceCriteria{
-				Cluster:                cluster,
-				Namespace:              namespace,
-				IncludeHealth:          false,
-				IncludeIstioResources:  false,
-				IncludeOnlyDefinitions: true,
-				ServiceSelector:        labels.Set(w.Labels).String(),
-			}
-			ss, err = in.businessLayer.Svc.GetServiceList(ctx, serviceCriteria)
-			if err != nil {
-				return nil, err
+
+		wg.Add(1) // per workload to load in parallel
+
+		go func(k string, w models.Workload) {
```

Comment:
To be honest, I'm a little bit afraid of this.  The number of apps in a namespace could be quite large, and so it could launch quite a high number of goroutines.  Each of those is going to fetch information. I realize that it's likely cached but I'm still a little nervous about the somewhat unbounded launch and whether it should maybe be chunked to control the number of concurrent funcs and the cache access.  I'm going to ask @nrfox to also comment...

Reply:
Yes this would create potentially a lot of goroutines because anywhere that calls `fetchNamespaceApps` would create `numberOfApps` goroutines. In a single request, `fetchNamespaceApps` could be called any number of times and there might be multiple requests in flight that each call `fetchNamespaceApps`. Moreover I don't think there's much to gain by parallelizing this. All of the information is already cached anyways so you're not saving any time. This would be like spinning up a separate goroutine to read each value out of a map or slice. You're better off just iterating over the map/slice.

If there was a perf degradation in the apps endpoint in the last release, were there any changes to `business/apps` in the last release?